### PR TITLE
Add the ability to clean up the dentry cache can be used to clean up resources when VFS umount.

### PR DESCRIPTION
**Pull request by zyfjeff**
1. In order to make the hot upgrade of virtiofs easy, VFS will save pseudo
inodes when umount for easy recovery. However, in the fuse scenario, if
umount does not remove the pseudo inode, it will cause an invalid
directory to be seen on the host, which is not friendly to users. So add
this option to control this behavior.

2. umount returns the parent's inode information, so that it is convenient
to call fuse's invali entry interface later to clarify the corresponding
dentry cache

3. Use the invalidate entry provided by the fuse kernel to clear the entry
cache, all code references the libfuse implementation.

4. Optimize log output and print in vfs umount scenario
      1. optimize the log of forget, for scenarios where forget cannot handle when the vfs submount is dropped by umount.
      2. remove the error log printed when write fuse fails, and hand it to the caller to process and print.

Signed-off-by: zyfjeff <zyfjeff@linux.alibaba.com>

ref: https://github.com/libfuse/libfuse/blob/f44214be6a2abb4c98f61790cae565c06bdb431c/lib/fuse_lowlevel.c#L2332

### CODE REVIEW

**Review by eryugey — APPROVED** [review]

File: src/api/vfs/mod.rs [hunk 1]
```diff
@@ -397,6 +404,9 @@ impl Vfs {
             .map_err(VfsError::PathWalk)?
             .ok_or_else(|| VfsError::NotFound(path.to_string()))?;
 
+        // Safe unwrap, because the inode must exists
+        let parent = self.root.get_parent_inode(inode).unwrap();
```
Comment (jiangliu) [inline src/api/vfs/mod.rs:408]: How about returning an error here?

File: src/api/pseudo_fs.rs [hunk 2]
```diff
@@ -235,6 +235,12 @@ impl PseudoFs {
         self.inodes.store(Arc::new(hashmap));
     }
 
+    pub fn get_parent_inode(&self, ino: u64) -> Option<u64> {
```
Comment (jiangliu) [inline src/api/pseudo_fs.rs:238]: The commit message doesn't match the change.

File: src/lib.rs [hunk 3]
```diff
@@ -76,6 +76,8 @@ pub enum Error {
     InvalidXattrSize((u32, usize)),
     /// Invalid message that the server cannot handle properly.
     InvalidMessage(io::Error),
+    /// Failed to writer buf to writer
```
Comment (jiangliu) [inline src/lib.rs:79]: Failed to write buffer to writer.

File: src/api/server/sync_io.rs [hunk 4]
```diff
@@ -19,10 +21,46 @@ use crate::abi::virtio_fs::{RemovemappingIn, RemovemappingOne, SetupmappingIn};
...
 };
+#[cfg(feature = "fusedev")]
+use crate::transport::FuseDevWriter;
 use crate::transport::{pagesize, FsCacheReqHandler, Reader, Writer};
 use crate::{bytes_to_cstr, encode_io_error_kind, BitmapSlice, Error, Result};
 
 impl<F: FileSystem + Sync> Server<F> {
+    #[cfg(feature = "fusedev")]
+    /// Use to send notify msg to kernel fuse
+    pub fn send_inval_entry_notify<S: BitmapSlice>(
```
Comment (jiangliu) [inline src/api/server/sync_io.rs:32]: `notify_inval_entry` or `send_inval_entry_notification`?

File: src/api/server/sync_io.rs [hunk 5]
```diff
@@ -19,10 +21,46 @@ use crate::abi::virtio_fs::{RemovemappingIn, RemovemappingOne, SetupmappingIn};
...
+        &self,
+        w: FuseDevWriter<'_, S>,
+        parent: u64,
+        name: &CStr,
```
Comment (jiangliu) [inline src/api/server/sync_io.rs:36]: Replace `&CStr` with `std::ffi::CStr`, the line 6-7 can be removed.

File: src/api/vfs/mod.rs (hunk 1 above)
Comment (zyfjeff) [reply src/api/vfs/mod.rs]: fixed

File: src/api/vfs/mod.rs [hunk 6]
```diff
@@ -406,7 +416,9 @@ impl Vfs {
                 // 1. they can be reused later on
                 // 2. during live upgrade, it is easier reconstruct pseudofs inodes since
                 //    we do not have to track pseudofs deletions
-                //self.root.evict_inode(inode);
+                if self.remove_pseudo_root {
```
[the remark this answers is not in the document]
Comment (zyfjeff) [reply src/api/vfs/mod.rs:430]: fixed

File: src/lib.rs (hunk 3 above)
Comment (zyfjeff) [reply src/lib.rs]: fixed

File: src/api/server/sync_io.rs [hunk 7]
```diff
@@ -19,10 +21,46 @@ use crate::abi::virtio_fs::{RemovemappingIn, RemovemappingOne, SetupmappingIn};
...
+    ) -> Result<usize> {
+        let mut writer = w;
+        // Safely to unwrap, offset 0 always large equal then buffer size
+        let mut buffer_writer = writer.split_at(0).unwrap();
```
[the remark this answers is not in the document]
Comment (zyfjeff) [reply src/api/server/sync_io.rs]: fixed

File: src/api/server/sync_io.rs (hunk 4 above)
Comment (zyfjeff) [reply src/api/server/sync_io.rs]: fixed

File: src/api/server/sync_io.rs (hunk 5 above)
Comment (zyfjeff) [reply src/api/server/sync_io.rs]: fixed

**Review by jiangliu — APPROVED** [review]

**Pull request merged**